// dbcontext/dist/classes/dbcontext.js
import fs from "fs";
import DbSet from "../classes/db-set.js";
import Wait from "../utils/wait.js";
class DbContext {
    options;
    path = "./db.json";
    intervalId;
    loaded = false;
    constructor(path = "./db.json", options) {
        this.options = {
            saveInterval: options?.saveInterval || 1000 * 5,
        };
        this.path = path;
        this.load();
        this.intervalId = setInterval(() => {
            this.save();
        }, this.options.saveInterval);
    }
    async WaitForLoad() {
        while (!this.loaded) {
            await Wait(100);
        }
    }
    async load() {
        if (!fs.existsSync(this.path))
            return;
        let data = await fs.promises.readFile(this.path, "utf8");
        let json = JSON.parse(data);
        for (let key in json) {
            this[key] = new DbSet(json[key].name, json[key].rows);
        }
        await Wait(1000);
        this.loaded = true;
    }
    async save() {
        let result = {};
        for (let key in this) {
            if (this[key] instanceof DbSet) {
                result[key] = this[key];
            }
        }
        await fs.promises.writeFile(this.path, JSON.stringify(result, null, 4));
    }
    async close() {
        clearInterval(this.intervalId);
        await this.save();
    }
}
export default DbContext;
